src/spec-configuration/containerFeaturesOrder.ts

```diff
@@ -3,133 +3,631 @@
+
+	if ((worklist.length + legacyWorklist.length) === 0) {
+		output.write('Zero length or undefined worklist.', LogLevel.Error);
+		return;
+	}
+
+	output.write(`${JSON.stringify(worklist, null, 2)}`, LogLevel.Trace);
+
+	// Sanity check
+	if (worklist.some(node => !node.featureSet) || legacyWorklist.some(node => !node.userFeatureId)) {
+		output.write('Feature dependency worklist contains one or more undefined entries.', LogLevel.Error);
+		throw new Error(`ERR: Failure resolving Feature dependencies.`);
+	}
+
+	output.write(`[raw worklist]: ${worklist.map(n => n.userFeatureId).join(', ')}`, LogLevel.Trace);
+
+	// For each node in the worklist, remove all 'soft-dependency' graph edges that are irrelevant
+	// i.e. the node is not a 'soft match' for any node in the worklist itself
+	for (let i = 0; i < worklist.length; i++) {
+		const node = worklist[i];
+		// reverse iterate
+		for (let j = node.installsAfter.length - 1; j >= 0; j--) {
+			const softDep = node.installsAfter[j];
+			if (!worklist.some(n => satisfiesSoftDependency(params, n, softDep))) {
+				output.write(`Soft-dependency '${softDep.userFeatureId}' is not required.  Removing from installation order...`, LogLevel.Info);
+				// Delete that soft-dependency
+				node.installsAfter.splice(j, 1);
+			}
+		}
+	}
+
+	output.write(`[worklist-without-dangling-soft-deps]: ${worklist.map(n => n.userFeatureId).join(', ')}`, LogLevel.Trace);
+	output.write('Starting round-based Feature install order calculation from worklist...', LogLevel.Trace);
+
+	const installationOrder: FNode[] = [];
+	while (worklist.length > 0) {
+		const round = worklist.filter(node =>
+			// If the node has no hard/soft dependencies, the node can always be installed.
+			(node.dependsOn.length === 0 && node.installsAfter.length === 0)
+			// OR, every hard-dependency (dependsOn) AND soft-dependency (installsAfter) has been satified in prior rounds
+			|| node.dependsOn.every(dep =>
+				installationOrder.some(installed => equals(params, installed, dep)))
+			&& node.installsAfter.every(dep =>
+				installationOrder.some(installed => satisfiesSoftDependency(params, installed, dep))));
+
+		output.write(`\n[round] ${round.map(r => r.userFeatureId).join(', ')}`, LogLevel.Trace);
+		if (round.length === 0) {
+			output.write('Circular dependency detected!', LogLevel.Error);
+			output.write(`Nodes remaining: ${worklist.map(n => n.userFeatureId!).join(', ')}`, LogLevel.Error);
+			return;
+		}
+
+		// Delete all nodes present in this round from the worklist.
+		worklist.splice(0, worklist.length, ...worklist.filter(node => !round.some(r => equals(params, r, node))));
+
+		// Sort rounds lexicographically by id.
+		round.sort((a, b) => comparesTo(params, a, b));
+
+		output.write(`[round-sort-id] ${round.map(r => r.userFeatureId).join(', ')}`, LogLevel.Trace);
+
+		// Final sort of rounds by priority
+		// Higher priority nodes are installed earlier.
+		round.sort((a, b) => b.roundPriority - a.roundPriority);
```

Comment:
I think `round` needs to be filtered down to the nodes with the highest `roundPriority` otherwise `overrideFeatureInstallOrder` doesn't have the global effect it is meant to have.

E.g., if B depends on A and C is independent and `overrideFeatureInstallOrder` is A, B, C, the above would install A and C in the first round and B in the second. If in each round you only install the highest priority nodes, the first round will install A, the second round will install B and only the third round will install C.

Reply:
I see, so the algorithm 

- From `dependsOn

Reply:
Ok, I think I understand the distinction.  I agree that overrideFeatureInstallOrder is only evaluated locally per round given my current implementation.

I will capture this example in a test and make the change.

Reply:
Fixed in https://github.com/devcontainers/cli/pull/530/commits/9b8eb0f111bc44c2466e5f0b4fc1661669fc5a85